cpp/src/merge/merge.cu

```diff
@@ -353,76 +354,154 @@ std::unique_ptr<column> column_merger::operator()<cudf::dictionary32>(column_vie
 
+table_ptr_type merge(cudf::table_view const& left_table,
+                     cudf::table_view const& right_table,
+                     std::vector<cudf::size_type> const& key_cols,
+                     std::vector<cudf::order> const& column_order,
+                     std::vector<cudf::null_order> const& null_precedence,
+                     rmm::mr::device_memory_resource* mr,
+                     cudaStream_t stream = 0) {
     //collect index columns for lhs, rhs, resp.
     //
     cudf::table_view index_left_view{left_table.select(key_cols)};
     cudf::table_view index_right_view{right_table.select(key_cols)};
-    bool nullable = cudf::has_nulls(index_left_view) || cudf::has_nulls(index_right_view);
+    bool const nullable = cudf::has_nulls(index_left_view) || cudf::has_nulls(index_right_view);
 
     //extract merged row order according to indices:
     //
     rmm::device_vector<index_type>
       merged_indices = generate_merged_indices(index_left_view, index_right_view, column_order, null_precedence, nullable);
 
     //create merged table:
-    //
-    std::vector<std::unique_ptr<column>> v_merged_cols;
-    v_merged_cols.reserve(n_cols);
+    //    
+    auto const n_cols = left_table.num_columns();
+    std::vector<std::unique_ptr<column>> merged_cols;
+    merged_cols.reserve(n_cols);
 
     column_merger merger{merged_indices, mr, stream};
+    transform(left_table.begin(), left_table.end(), 
+              right_table.begin(), 
+              std::back_inserter(merged_cols),
+              [&](auto const& left_col, auto const& right_col){
+                return cudf::experimental::type_dispatcher(left_col.type(),
+                                                           merger,
+                                                           left_col,
+                                                           right_col);
+              });
+
+    return std::make_unique<cudf::experimental::table>(std::move(merged_cols));
+}
 
-    for(auto i=0;i<n_cols;++i)
-      {
-        const auto& left_col = left_table.column(i);
-        const auto& right_col= right_table.column(i);
-
-        auto merged = cudf::experimental::type_dispatcher(left_col.type(),
-                                                          merger,
-                                                          left_col,
-                                                          right_col);
-        v_merged_cols.emplace_back(std::move(merged));
-      }
+struct merge_queue_item {
+  table_view view_;
+  table_ptr_type table_;
+  // Priority is a separate member to ensure that moving from an object 
+  // does not change its priority (which would ruin the queue invariant)
+  cudf::size_type priority_ = 0;
+ 
+  void set_priority() {
+    // Smallest tables have the highest priority
+    priority_ = -view_.num_rows();
+  }
+ 
+  merge_queue_item(table_view const& view, table_ptr_type&& table):
+  view_(view), 
```

Comment:
Minor stylistic preference, here, but can you indent the initializations so they are easily distinguishable from the constructor name?  Also, suggest using  {} instead of ().

```
merge_queue_item(table_view const& view, table_ptr_type&& table):
  view{view}, table{std::move(table)} {
```